### PR TITLE
fix(datepicker): validate using ng-required attribute

### PR DESCRIPTION
**Pull request by onenive**
This is a solution for the the issue (https://github.com/angular-ui/bootstrap/issues/3862), where date is always required for datepicker, even when ng-required is set to false.

(todo: still need to add tests)

### CODE REVIEW

**Comment by onenive** [comment]
@wesleycho, moving the discussion to this PR: 

In response to your comments, I had some questions. Would the proper fix be to:
 
1) log the error: "Datepicker directive: "ng-model" value must be a Date object..." regardless of what ng-required is set as
2) remove ngModelCtrl.$setValidity('date', ...) from this.render() and move all the validation logic to ngModelCtrl.$validators.date

**Comment by wesleycho** [comment]
Correct on both counts

**Comment by onenive** [comment]
does this look good?

File: src/datepicker/datepicker.js [hunk 1]
```diff
@@ -616,6 +615,11 @@ function ($compile, $parse, $document, $position, dateFilter, dateParser, datepi
 
       function validator(modelValue, viewValue) {
         var value = modelValue || viewValue;
+
+        if (!attrs.ngRequired && !value) {
```
Comment (wesleycho) [inline src/datepicker/datepicker.js]: Is this even necessary? Does this run if ngRequired is falsy?

Comment (onenive) [inline src/datepicker/datepicker.js]: Yes, it works even if ngRequired is falsy. If ngRequired is falsy (i.e. anything other than ng-required=true), then the datepicker is not required. So, it considers an empty value valid in that case.

Comment (wesleycho) [inline src/datepicker/datepicker.js]: I wasn't saying just working - I was wondering whether the validator even runs if `!attrs.ngRequired`. If it doesn't run, it wouldn't be necessary, but if it does, then this line is fine.

Comment (onenive) [inline src/datepicker/datepicker.js]: The validator does run if !attrs.ngRequired

**Comment by wesleycho** [comment]
This LGTM, I will do a full check on Sunday and hopefully will merge it in shortly after.

**Comment by onenive** [comment]
hey, is there any update on this?

**Comment by wesleycho** [comment]
Sorry, haven't had the chance to check this out yet, as it requires a little more setup/evaluation than the other stuff I have looked at so far.

I do intend on reviewing this thoroughly this week though.

**Comment by wesleycho** [comment]
Looks good [here](http://plnkr.co/edit/gIp1ixngsjx8pEp6jU7N?p=preview).

**Pull request closed without merging**